REMS-121: Prepoulated User

tstrass opened the pull request:
Add a script for populating PIS DB with a user and call the script from the Dockerfile so it's baked into the Docker
image.

tstrass committed:
Add a script for populating PIS DB with a user

Also call the script from the Dockerfile so it's baked into the Docker
image.

smalho01 commented on backend/populatedb.js:
```diff
@@ -0,0 +1,21 @@
+mongoose.connect(mongoConnString, {useNewUrlParser: true , useUnifiedTopology: true});
+
+const user = new User({
+  name: 'alice',
```
can we change the name and correspondingly email from Alice to a different persona so that its clear during the demo that the pharmacist is separate from the prescriber in the workflow. The chosen name isn't as important so long as it's different. Also we should set a full name here rather than just a first name

tstrass replied:
Changed https://github.com/mcode/pharmacy-information-system/pull/11/commits/636459aead979959170b0d5b5d18c77810bab835

smalho01 commented on backend/populatedb.js:
```diff
@@ -0,0 +1,21 @@
+
+// If we don't disable validation, mongoose-unique-validator will erroneously throw a ValidationError
+// on the User email unique constraint
+user.save({ validateBeforeSave: false })
```
this is interesting, wouldn't it only complain about validation on the unique email constraint if an entry with that email already exists? I haven't tried running this yet, will do so soon,  but it seems like this save function is executed every time the container starts up based on the docker file cmd and thus resulting in multiple repeat entries into the users mongo collection with validation disabled. 

A better approach would be to check if the user already exists in the db with a find, if it does exists then do nothing, if it doesn't exist then and only then create a new entry. This will make it so that you don't need to disable validation.

tstrass replied:
>wouldn't it only complain about validation on the unique email constraint if an entry with that email already exists?

I was getting the validation error even when there were zero existing users (hence the "erroneously" in my comment). I couldn't figure out why it was happening, but if you want to try removing the `validateBeforeSave: false` in your testing, please let me know if you get the same result.

>it seems like this save function is executed every time the container starts up based on the docker file cmd and thus resulting in multiple repeat entries into the users mongo collection with validation disabled.

Good point. I'll add a check for existing user. But I still think I'll have to disable validation due to the issue I described.

smalho01 replied:
I just tested this locally and was able to save the login credentials with and without the validation when the db is empty. Trying to run it again though when there is a duplicate email already in the db gives an error in both cases with the validation disabled or removed, though each case gives a different error. I didn't test within docker, but it should work just the same as running locally in theory. I think having a check for an existing user will fix the error in both cases though. Make sure you're looking at the Users table and not the DoctorUsers table as the latter isn't really used I don't think and see if you're able to save without validation. If for some reason that's giving you issue, we can schedule some time to debug it

tstrass replied:
I did check that the database was empty when I got the seemingly erroneous duplicate email error, but after doing this refactor, I cannot reproduce anymore (🤷), so I removed the disabling of validation.

Note, however, that I decided not to add a check for if the user exists already because I don't think we really get any benefit from doing so; if the user already exists when the container starts up, it will simply throw the correct validation error and continue starting up the application. https://github.com/mcode/pharmacy-information-system/pull/11/commits/636459aead979959170b0d5b5d18c77810bab835

smalho01 commented on backend/populatedb.js:
```diff
@@ -0,0 +1,21 @@
+  contact: '555-555-5555',
+  nic: '1234',
+  email: 'alice@example.com' ,
+  password: '$2a$10$CMRf90.VvXfd9CN04duIi.QvtO.US1Gcdzs3aRJsbWztKBstm8wgy',
```
what is the unencrypted form of this password? Should we use bycrypt.hash() here like in backend/routes/user.js?

tstrass replied:
https://github.com/mcode/pharmacy-information-system/pull/11/commits/636459aead979959170b0d5b5d18c77810bab835

smalho01 commented on Dockerfile:
```diff
@@ -7,4 +7,4 @@ RUN npm install
 EXPOSE 3010
 EXPOSE 4200
-CMD npm run start
+CMD node backend/populatedb.js && npm run start
```
this should also be added to Dockerfile.dev for the dev env - that docker file uses a shell script that the `node backend/populatedb.js` command should be added to

smalho01 replied:
This still needs to be added to the dev environment, but otherwise this looks good

tstrass replied:
~https://github.com/mcode/pharmacy-information-system/pull/11/commits/59dd970c9fad12f5d3544e769aaac8e0e5333e19~ https://github.com/mcode/pharmacy-information-system/pull/11/commits/8a838aa672e9782f4e0821dc30a642aab6862b9c also removed some whitespace

tstrass committed:
Modify prepoulated user data

- Change email and name
- Generate a hashed password from plaintext
- Remove unneeded disabling of validation

tstrass committed:
Add prepoulated db data to dev docker image

Also remove some extraneous whitespace in the dockerRunnerDev.sh script
